sdk/typescript/src/client/ika-transaction.ts

```diff
@@ -0,0 +1,1722 @@
+	async signWithImportedDWalletPublic({
+		dWallet,
+		importedKeyMessageApproval,
+		verifiedPresignCap,
+		presign,
+		hashScheme,
+		message,
+		ikaCoin,
+		suiCoin,
+	}: {
+		dWallet: DWallet;
+		importedKeyMessageApproval: TransactionObjectArgument;
+		verifiedPresignCap: TransactionObjectArgument;
+		presign: Presign;
+		hashScheme: Hash;
+		message: Uint8Array;
+		ikaCoin: TransactionObjectArgument;
+		suiCoin: TransactionObjectArgument;
+	}) {
+		if (!dWallet.state.Active?.public_output) {
+			throw new Error('DWallet public output is not set');
+		}
+
+		if (!presign.state.Completed?.presign) {
+			throw new Error('Presign is not completed');
+		}
+
+		if (!dWallet.public_user_secret_key_share) {
+			throw new Error('DWallet public user secret key share is not set');
+		}
+
+		const publicParameters = await this.ikaClient.getProtocolPublicParameters();
+
+		coordinatorTx.requestImportedKeySign(
+			this.ikaClient.ikaConfig,
+			this.getCoordinatorObjectRef(),
+			verifiedPresignCap,
+			importedKeyMessageApproval,
+			createUserSignMessage(
+				publicParameters,
+				dWallet,
+				Uint8Array.from(dWallet.public_user_secret_key_share),
+				Uint8Array.from(presign.state.Completed?.presign),
+				message,
+				hashScheme,
+			),
+			this.createSessionIdentifier(),
+			ikaCoin,
+			suiCoin,
+			this.transaction,
+		);
+	}
+
+	/**
+	 * Transfer an encrypted user share from the current user to another address.
+	 * This re-encrypts the user's share with the destination address's encryption key.
+	 *
+	 * @param params - The parameters for transferring encrypted user share
+	 * @param params.dWallet - The DWallet whose user share is being transferred
+	 * @param params.destinationSuiAddress - The Sui address that will receive the re-encrypted share
+	 * @param params.sourceEncryptedUserSecretKeyShare - The current user's encrypted secret key share
+	 * @param params.ikaCoin - The IKA coin object to use for transaction fees
+	 * @param params.suiCoin - The SUI coin object to use for gas fees
+	 * @returns Promise resolving to the updated IkaTransaction instance
+	 * @throws {Error} If user share encryption keys are not set
+	 */
+	async transferUserShare({
+		dWallet,
+		destinationSuiAddress,
+		sourceEncryptedUserSecretKeyShare,
+		ikaCoin,
+		suiCoin,
+	}: {
+		dWallet: DWallet;
+		destinationSuiAddress: string;
+		sourceEncryptedUserSecretKeyShare: EncryptedUserSecretKeyShare;
+		ikaCoin: TransactionObjectArgument;
+		suiCoin: TransactionObjectArgument;
+	}) {
+		if (!this.userShareEncryptionKeys) {
+			throw new Error('User share encryption keys are not set');
+		}
+
+		const publicParameters = await this.ikaClient.getProtocolPublicParameters();
+
+		const destinationEncryptionKeyObj =
+			await this.ikaClient.getActiveEncryptionKey(destinationSuiAddress);
+
+		coordinatorTx.requestReEncryptUserShareFor(
+			this.ikaClient.ikaConfig,
+			this.getCoordinatorObjectRef(),
+			dWallet.id.id,
+			destinationSuiAddress,
+			encryptSecretShare(
```

Comment:
I see, here you call a crypto function too. Let's think if we can separate responsibilities. I understand why you would want this here since you first decrypt the secret share.